warc_metadata_sidecar.py

```diff
@@ -0,0 +1,261 @@
+                break
+        return buffer
+
+    def print_matches(self, fullname, matches, delta_t, matchtype=''):
+        """This method prints out information for each match.
+
+        We override the method in order to save the mimetype and puid to the fido object.
+        """
+        # https://github.com/openpreserve/fido/blob/master/fido/fido.py#L272
+        if not matches:
+            self.puid = None
+            self.mime = None
+            return
+        match_format, _ = matches[0]
+        mime = match_format.find('mime')
+        self.mime = mime.text if mime is not None else None
+        self.puid = self.get_puid(match_format)
+
+
+def find_mime_and_puid(fido, payload, url):
+    """Find the mimetype and preservation identifier using fido."""
+    fido.identify_stream(payload, url)
+
+
+def find_character_set(bytes_payload):
+    """Find the character set of the payload using chardet."""
+    result = chardet.detect(bytes_payload)
+    result_dict = {'encoding': result['encoding'],
+                   'confidence': result['confidence']
+                   }
+    return result_dict
+
+
+def find_language(bytes_payload):
+    """Find the language of the payload using pycld2."""
+    is_reliable, text_bytes_found, details = cld2.detect(bytes_payload.decode('utf-8', 'replace'),
+                                                         bestEffort=True)
+    new_list = []
+    # 'details' seems to always return 3, if the language is 'Unknown' we don't need to list it.
+    for item in details:
+        if 'Unknown' not in item:
+            new_list.append({
+                'name': item[0],
+                'code': item[1],
+                'text-covered': item[2],
+                'score': item[3]})
+    if len(new_list):
+        lang_cld = {'reliable': is_reliable,
+                    'text-bytes': text_bytes_found,
+                    'languages': new_list}
+        return lang_cld
+    else:
+        return None
+
+
+def create_warcinfo_payload(payload, operator, publisher, new_file, version):
+    payload.seek(0)
+    info_payload = payload.read()
+    warcinfo = info_payload.decode("utf-8")
+    warc_list = re.split('\r\n|: ', warcinfo)
+    hostname = socket.gethostname()
+    warc_doc = 'http://bibnum.bnf.fr/WARC/WARC_ISO_28500_version1_latestdraft.pdf'
+    warcinfo_payload = {'software': 'warc-metadata-sidecar/' + version,
+                        'hostname': hostname,
+                        'ip': socket.gethostbyname(hostname),
+                        'conformsTo': warc_doc,
+                        'description': 'WARC metdata sidecar for ' + new_file,
+                        'publisher': publisher}
+    if operator:
+        warcinfo_payload['operator'] = operator
+    if 'isPartOf' in warc_list:
+        list_index = warc_list.index('isPartOf')
+        is_part_of = warc_list[list_index + 1]
+        warcinfo_payload['isPartOf'] = is_part_of
```

Comment:
Rather than how you are creating the warc_list and searching it to find isPartOf, we could perhaps remove the lines:
```
    info_payload = payload.read()
    warcinfo = info_payload.decode("utf-8")
    warc_list = re.split('\r\n|: ', warcinfo)
```
and use a regex to look for it instead of the `if 'isPartOf' in warc_list` block:
```
    # Check if the original warcinfo contains an isPartOf field. 
    is_part_of = re.search(r'\bisPartOf: +([^\r\n]+)', payload.read().decode('utf-8'))
    if is_part_of is not None:
            warcinfo_payload['isPartOf'] = is_part_of.group(1)
```
(exact code here was not run and could have typos)

Reply:
@ldko Did I understand that we can get rid of isPartOf? That would mean that we can create the warcinfo record before the for loop.